### PR TITLE
feat(migrate diff): add -o --output option

### PR DESCRIPTION
**Pull request by Jolg42**
Using `-o` or `--output` as they are well known options, see https://clig.dev/#arguments-and-flags

Since the Schema engine does not return the content of the diff, I had to intercept it.
To achieve that, it was best, and something we wanted to do a few years ago, to use `process.stdout`, and since many tests rely on the output the git diff is quite large.

Note: I did not want to use our Jest mock for process.stdout because it captures everything, which makes it very ugly.
Example: `DEBUG=prisma*` is captured, making it very hard to debug anything 🙈 

~~TODO: cherry pick this to main later~~

### CODE REVIEW

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 1]
```diff
@@ -18,42 +18,14 @@ const describeIf = (condition: boolean) => (condition ? describe : describe.skip
...
-        PRAGMA foreign_keys=off;
-        DROP TABLE "Post";
-        PRAGMA foreign_keys=on;
-
-        -- DropTable
-        PRAGMA foreign_keys=off;
-        DROP TABLE "User";
-        PRAGMA foreign_keys=on;
-      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:28]: Please replace `ctx.mocked['console.info']` with `ctx.mocked['process.stdout.write']`, so we don't lose track of the SQL script in the snapshot.

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 2]
```diff
@@ -62,26 +34,7 @@ describe('migrate diff', () => {
...
-            "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
-            "email" TEXT NOT NULL,
-            "count1" INTEGER NOT NULL,
-            "name" TEXT
-        );
-
-        -- CreateIndex
-        CREATE UNIQUE INDEX "User_email_key" ON "User"("email" ASC);
-      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:37]: Please replace `ctx.mocked['console.info']` with `ctx.mocked['process.stdout.write']`, so we don't lose track of the SQL script in the snapshot.

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 3]
```diff
@@ -100,17 +53,7 @@ describe('migrate diff', () => {
...
-        PRAGMA foreign_keys=off;
-        DROP TABLE "Post";
-        PRAGMA foreign_keys=on;
-
-        -- DropTable
-        PRAGMA foreign_keys=off;
-        DROP TABLE "User";
-        PRAGMA foreign_keys=on;
-      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:56]: Please replace `ctx.mocked['console.info']` with `ctx.mocked['process.stdout.write']`, so we don't lose track of the SQL script in the snapshot.

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 4]
```diff
@@ -129,26 +72,7 @@ describe('migrate diff', () => {
...
-            "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
-            "email" TEXT NOT NULL,
-            "count1" INTEGER NOT NULL,
-            "name" TEXT
-        );
-
-        -- CreateIndex
-        CREATE UNIQUE INDEX "User_email_key" ON "User"("email" ASC);
-      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:75]: Please replace `ctx.mocked['console.info']` with `ctx.mocked['process.stdout.write']`, so we don't lose track of the SQL script in the snapshot.

File: packages/migrate/src/commands/MigrateDiff.ts [hunk 5]
```diff
@@ -296,6 +309,14 @@ ${bold('Examples')}
       migrate.stop()
     }
 
+    // Write output to file if --output is defined
+    if (isOutputDefined) {
+      captureStdout.stopCapture()
+      const diffOutput = captureStdout.getCapturedText()
+      captureStdout.clearCaptureText()
+      await fs.writeAsync(outputPath!, diffOutput.join('\n'))
```
Comment (Jolg42) [inline packages/migrate/src/commands/MigrateDiff.ts:317]: I did not add any error handling here, if we want it can be done later

File: packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap [hunk 6]
```diff
@@ -1,84 +1,50 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`migrate diff generic should fail when --from-local-d1 is used with --shadow-database-url 1`] = `
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap:3]: What happened to this snapshot?

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 7]
```diff
@@ -197,7 +134,7 @@ describe('migrate diff', () => {
         'file:dev.db',
         '--script',
       ])
-      await expect(result).rejects.toMatchSnapshot()
+      await expect(result).rejects.toThrow()
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:137]: Why this change?

File: packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap (hunk 6 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap:3]: It's not interesting to capture the full output here, it makes it hard to change the help output because so many snapshots rely on it
So I changed this to check that it throws.

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 7 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:137]: Like said above, relying on the full output for failures makes it hard to edit the help output of the command.
It's better to only match part of the output if we want

Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts]: See new commit [`07a9aa7` (#23715)](https://github.com/prisma/prisma/pull/23715/commits/07a9aa769b13acc04b34f392dff6e1e7a5fb414d)

File: packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap [hunk 8]
```diff
@@ -1,84 +1,50 @@
...
-  --script                 Render a SQL script to stdout instead of the default human readable summary (not supported on MongoDB)
-  --exit-code              Change the exit code behavior to signal if the diff is empty or not (Empty: 0, Error: 1, Not empty: 2). Default behavior is Success: 0, Error: 1.
+-- DropTable
+PRAGMA foreign_keys=off;
+DROP TABLE "User";
+PRAGMA foreign_keys=on;
 
 `;
 
-exports[`migrate diff generic should fail when --to-local-d1 is used with --shadow-database-url 1`] = `
```
[the remark this answers is not in the document]
Comment (Jolg42) [reply packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap:44]: See new commit [`07a9aa7` (#23715)](https://github.com/prisma/prisma/pull/23715/commits/07a9aa769b13acc04b34f392dff6e1e7a5fb414d)

File: packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap (hunk 6 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap:3]: See new commit [`07a9aa7` (#23715)](https://github.com/prisma/prisma/pull/23715/commits/07a9aa769b13acc04b34f392dff6e1e7a5fb414d)

**Review by jkomyno — APPROVED** [review]
